src/mount/lite.clj

```diff
@@ -18,17 +18,36 @@
 
 ;;; The state protocol implementation.
 
-(defonce ^:private itl (InheritableThreadLocal.))
+(defn new-session
+  []
+  (gensym "mount-session-"))
+
+(defonce ^:private default-session (new-session))
+
+(defonce itl
+  (proxy [InheritableThreadLocal] []
+    (initialValue []
+      default-session)))
+
+(defonce ^:dynamic *session* default-session)
+
+(defn current-session
+  []
+  (or (.get ^InheritableThreadLocal itl) *session*))
```

Comment:
Combining the approaches of using the InheritableThreadLocal and a binding so we can fall back to the `*session*` if the InheritableThreadLocal is missing.